### DISCUSSION TITLE
NFSv4 home options and nolock

### DISCUSSION THREAD

**Discussion by vsuojanen** [ltsp/ltsp#762] [Q&A]
When mounting user homes over NFS the mount options in server ltsp.conf show different when I boot a chroot image (debootsrap focal + ubuntu-desktop) compared to when I boot the server root (/) image (Ubuntu Focal 20.04.5 Desktop LTS). 

With the ubuntu focal chroot image NFS home mount options are working better for the client programs but would like to have the server root (/) as  the client template. 

>The problem when running with server root image is that program can only run once and after that start to give "Error in file locking Bad file descriptor".

I checked also Ubuntu Mate 20.04.5 with server root (/) image but NFS home mount options didn't work that well. The problem when running with server root image is again that program can only run once and after that start to give "Error in file locking Bad file descriptor".  

The only thing I came up looking so far was checking nfs nolock option and comparing differences in the running clients between chroot and chrootless image. 

The options in server ltsp.conf:

> grep nfs /etc/ltsp/ltsp.conf
> HOME_DIR=/home/nfs
> FSTAB_HOME="server:/home/nfs /home nfs defaults,nolock 0 0"
> \# FSTAB_HOME="server:/home /home nfs defaults,nolock 0 0"
> 

This is how the nfs options are running from the chroot maintained client image:

> grep nfs /proc/mounts 
> 10.60.207.9:/home/nfs /home nfs rw,relatime,vers=3,rsize=262144,wsize=262144,namlen=255,hard,nolock,proto=tcp,port=2049,timeo=600,retrans=3,sec=sys,local_lock=all,addr=10.60.207.9 0 0
> grep nfs /proc/self/mountinfo 
> 477 35 0:56 / /home rw,relatime shared:264 - nfs 10.60.207.9:/home/nfs rw,vers=3,rsize=262144,wsize=262144,namlen=255,hard,nolock,proto=tcp,port=2049,timeo=600,retrans=3,sec=sys,local_lock=all,addr=10.60.207.9
> cat /proc/cmdline 
> vmlinuz initrd=ltsp.img initrd=initrd.img root=/dev/nfs nfsroot=10.60.207.9:/srv/ltsp ltsp.image=images/focal.img loop.max_part=9 BOOTIF=01-00-16-3e-e7-f0-00 

> grep root /proc/mounts 
> /dev/loop0 /run/initramfs/ltsp/0/looproot squashfs ro,relatime 0 0
> /run/initramfs/ltsp / overlay rw,relatime,lowerdir=/run/initramfs/ltsp/0/looproot,upperdir=/run/initramfs/ltsp/0/up,workdir=/run/initramfs/ltsp/0/work 0 0
> grep root /proc/self/mountinfo 
> 32 31 7:0 / /run/initramfs/ltsp/0/looproot ro,relatime shared:7 - squashfs /dev/loop0 ro
> 35 1 0:28 / / rw,relatime shared:1 - overlay /run/initramfs/ltsp rw,lowerdir=/run/initramfs/ltsp/0/looproot,upperdir=/run/initramfs/ltsp/0/up,workdir=/run/initramfs/ltsp/0/work

The chroot maintained image client kernel:

> uname -r
> 5.4.0-26-generic
> 
> $ cat /etc/os-release 
> NAME="Ubuntu"
> VERSION="20.04 LTS (Focal Fossa)"
> ID=ubuntu
> ID_LIKE=debian
> PRETTY_NAME="Ubuntu 20.04 LTS"
> VERSION_ID="20.04"
> HOME_URL="https://www.ubuntu.com/"
> SUPPORT_URL="https://help.ubuntu.com/"
> BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
> PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
> VERSION_CODENAME=focal
> UBUNTU_CODENAME=focal`

This is how it look when booting from the chrootless image:

> grep nfs /proc/mounts 
> server:/home/nfs /home nfs4 rw,relatime,vers=4.2,rsize=262144,wsize=262144,namlen=255,hard,proto=tcp,timeo=600,retrans=2,sec=sys,clientaddr=10.60.207.158,local_lock=none,addr=10.60.207.9 0 0
> grep nfs /proc/self/mountinfo 
> 837 35 0:60 / /home rw,relatime shared:465 - nfs4 server:/home/nfs rw,vers=4.2,rsize=262144,wsize=262144,namlen=255,hard,proto=tcp,timeo=600,retrans=2,sec=sys,clientaddr=10.60.207.158,local_lock=none,addr=10.60.207.9
> 
> grep root /proc/mounts 
> /dev/loop0 /run/initramfs/ltsp/0/looproot squashfs ro,relatime,errors=continue 0 0
> /run/initramfs/ltsp / overlay rw,relatime,lowerdir=/run/initramfs/ltsp/0/looproot,upperdir=/run/initramfs/ltsp/0/up,workdir=/run/initramfs/ltsp/0/work,xino=off 0 0
> grep root /proc/self/mountinfo 
> 32 31 7:0 / /run/initramfs/ltsp/0/looproot ro,relatime shared:7 - squashfs /dev/loop0 ro,errors=continue
> 35 1 0:29 / / rw,relatime shared:1 - overlay /run/initramfs/ltsp rw,lowerdir=/run/initramfs/ltsp/0/looproot,upperdir=/run/initramfs/ltsp/0/up,workdir=/run/initramfs/ltsp/0/work,xino=off
> 
> cat /proc/cmdline 
> vmlinuz initrd=ltsp.img initrd=initrd.img root=/dev/nfs nfsroot=10.60.207.9:/srv/ltsp ltsp.image=images/x86_64.img loop.max_part=9 BOOTIF=01-00-16-3e-e7-f0-00

The server and client OS kernel:

> uname -r
> 5.15.0-50-generic
> 
> \# cat /etc/os-release 
> NAME="Ubuntu"
> VERSION="20.04.5 LTS (Focal Fossa)"
> ID=ubuntu
> ID_LIKE=debian
> PRETTY_NAME="Ubuntu 20.04.5 LTS"
> VERSION_ID="20.04"
> HOME_URL="https://www.ubuntu.com/"
> SUPPORT_URL="https://help.ubuntu.com/"
> BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
> PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
> VERSION_CODENAME=focal
> UBUNTU_CODENAME=focal

When booting chrootless client I can locate /usr/sbin/mount.nfs4:

> ls -ltr /usr/sbin/mount.nfs*
> lrwxrwxrwx 1 root root      9 touko  24  2021 /usr/sbin/mount.nfs4 -> mount.nfs
> -rwsr-xr-x 1 root root 121688 touko  24  2021 /usr/sbin/mount.nfs

When booting the chroot image I've not found /usr/sbin/mount.nfs. Though it's using "the old NFSv3" like expected?

> /usr/sbin/mount.nfs -> /usr/lib/klibc/bin/nfsmount

Any suggestions? Should the  FSTAB_HOME be different for NFSv4? or should I revert somehow to the NFSv3 in the server (chrootless) client image

**Comment by alkisg** [comment]
I don't have a lot of time to read this in depth, but two hints are:

- You can use `FSTAB_HOME="server:/home/nfs /home nfs defaults,vers=3 0 0"` to force a specific NFS version, assuming of course that you have installed the `nfs-common` package in the image.
- You should be trying with the same kernel version; `dpkg -l | grep linux` will tell you which version you're using, and `sudo apt install linux-generic-hwe-22.04` or `sudo apt install linux-generic` will install the 5.15 or 5.4 kernels respectively.

**Reply by vsuojanen** [reply]
That works forcing NFSv3 with the vers=3 option in FSTAB_HOME (needed that here for chrootless client in Ubuntu Desktop 20.04.5). It did not have to change from 5.15 to 5.4 kernel

> FSTAB_HOME="server:/home/nfs /home nfs defaults,vers=3,nolock 0 0"


> grep nfs /proc/mounts 
> server:/home/nfs /home nfs rw,relatime,vers=3,rsize=262144,wsize=262144,namlen=255,hard,nolock,proto=tcp,timeo=600,retrans=2,sec=sys,mountaddr=10.60.207.9,mountvers=3,mountport=55314,mountproto=udp,local_lock=all,addr=10.60.207.9 0 0

Going back to the question above,
Which NFSv3 or NFSv4 configuration is suggested to be used with Ubuntu Desktop 20.04.5 when configuring /home over NFS?

I somehow lucky first followed the [Chroots wiki page](https://github.com/ltsp/ltsp/wiki/chroots) and then the recommended chrootless client installation.(Server was installed using Ubuntu Desktop 20.04.5, also tested chrootless client installation using server Ubuntu Mate 20.04 LTS) 

When booting the chroot image "the old NFSv3" is used, LTSP client did not find /usr/sbin/mount.nfs and used

>     /usr/sbin/mount.nfs -> /usr/lib/klibc/bin/nfsmount

If I try to follow the recommended chrootless client installation I get  usr/sbin/mount.nfs4:

>     ls -ltr /usr/sbin/mount.nfs*
>     lrwxrwxrwx 1 root root 9 touko 24 2021 /usr/sbin/mount.nfs4 -> mount.nfs
>     -rwsr-xr-x 1 root root 121688 touko 24 2021 /usr/sbin/mount.nfs

That nfs4 did not work without forcing to NFS version 3
edit: (program in client start to give Error in file locking Bad file descriptor)
